src/canopy/llm/cohere.py

```diff
@@ -0,0 +1,327 @@
+            api_key: Your Cohere API key. Defaults to None (uses the "CO_API_KEY" environment variable).
+            base_url: The base URL to use for the Cohere API. Defaults to None (uses the "CO_API_URL" environment variable if set, otherwise use default Cohere API URL).
+            **kwargs: Generation default parameters to use for each request. See https://platform.openai.com/docs/api-reference/chat/create
+                    For example, you can set the temperature, p, etc
+                    These params can be overridden by passing a `model_params` argument to the `chat_completion` methods.
+        """  # noqa: E501
+        super().__init__(model_name)
+        self._client = cohere.Client(api_key, api_url=base_url)
+        self.ignore_unrecognized_params = ignore_unrecognized_params
+        self.default_model_params = kwargs
+
+    def chat_completion(self,
+                        system_prompt: str,
+                        chat_history: Messages,
+                        context: Optional[Context] = None,
+                        *,
+                        stream: bool = False,
+                        max_tokens: Optional[int] = None,
+                        model_params: Optional[dict] = None,
+                        ) -> Union[ChatResponse, Iterable[StreamingChatChunk]]:
+        """
+        Chat completion using the Cohere API.
+
+        Note: this function is wrapped in a retry decorator to handle transient errors.
+
+        Args:
+            system_prompt: The system prompt to use for the chat completion (preamble).
+            chat_history: Messages (chat history) to send to the model.
+            context: Knowledge base context to use for the chat completion. Defaults to None (no context).
+            stream: Whether to stream the response or not.
+            max_tokens: Maximum number of tokens to generate. Defaults to None (generates until stop sequence or until hitting max context size).
+            model_params: Model parameters to use for this request. Defaults to None (uses the default model parameters).
+                          Dictonary of parametrs to override the default model parameters if set on initialization.
+                          For example, you can pass: {"temperature": 0.9, "top_p": 1.0} to override the default temperature and top_p.
+                          see: https://platform.openai.com/docs/api-reference/chat/create
+        Returns:
+            ChatResponse or StreamingChatChunk
+
+        Usage:
+            >>> from canopy.llm import OpenAILLM
+            >>> from canopy.models.data_models import UserMessage
+            >>> llm = CohereLLM()
+            >>> messages = [UserMessage(content="Hello! How are you?")]
+            >>> result = llm.chat_completion(messages)
+            >>> print(result.choices[0].message.content)
+            "I'm good, how are you?"
+        """  # noqa: E501
+        model_params_dict: Dict[str, Any] = deepcopy(self.default_model_params)
```

Comment:
May be incorrect here if already tried, but don't think a `deepcopy` is needed here

Reply:
@giannis2two I would suggest we keep the `deepcopy`. I don't think it's an issue. Also, that's how they do it in the `OpenAILLM.chat_completion` method - with the deep copy.

The reason is because of how the model params are merged. One set is passed to the `CohereLLM.__init__` method (which come from the YAML config file and apply to all chat completion calls), and then there are separate model params passed to the `CohereLLM.chat_completion` method. Those two sets of model params get merged, and if it didn't do the deep copy here, there would be risk of updating values in the `self.self.default_model_params` dict with values from `model_params` passed to the chat completion method. There's certainly other ways of merging the values of the two dicts together without using `deepcopy`, but that is the method they chose to merge the default model params and the ones passed on each chat completion call in the base `OpenAILLM.chat_completion` method, and I'm doing it the same way here.